Add blog post for v1.24 maxUnavailable for StatefulSet (#32638)
kubernetes/website: content/en/blog/_posts/2022-03-29-maxunavailable-for-statefulset.md

---
layout: blog

> title: 'Kubernetes 1.24: MaxUnavailable for StatefulSet'

Reviewer: In the API, the field is capitalized as `maxUnavailable`.

How about:
```suggestion
title: 'Kubernetes 1.24: Max Unavailable Replicas for StatefulSet'
```

> date: 2022-03-29

Reviewer 2: ```suggestion
date: 2022-05-27
```

slug: maxunavailable-for-statefulset
---

**Author:** Mayank Kumar (Salesforce)

> Kubernetes [StatefulSets](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/), since their introduction in 

Reviewer: ```suggestion
Kubernetes [StatefulSets](/docs/concepts/workloads/controllers/statefulset/), since their introduction in 
```

1.5 and becoming stable in 1.9, have been widely used to run stateful applications. They provide stable pod identity, persistent
per pod storage and ordered graceful deployment, scaling and rolling updates. You can think of StatefulSet as the atomic building
block for running complex stateful applications. As the use of Kubernetes has grown, so has the number of scenarios requiring
StatefulSets. Many of these scenarios, require faster rolling update then the currently supported one pod at a time, in

> OrderedReady PodManagementPolicy.

Reviewer: ```suggestion
the case where you're using the `OrderedReady` Pod management policy for a StatefulSet.
```

> Example:
> 1: StatefulSet is being used to model a multi instance cache based application where the size of the cache is large and requires

Reviewer: ```suggestion
Here are some examples:

1. I am using a StatefulSet to orchestrate a multi instance, cache-based application where the size of the cache is large. The cache starts cold and requires
```

   some siginificant amount of time before the container can start. There could be more initial startup tasks that are required. A
   RollingUpdate on this StatefulSet would take a lot of time before the application is fully updated. If the StatefulSet supported
   updating more than one pod at a time, it would result in a much faster update.

> 2: My Stateful application is composed of leaders and followers or one writer and multiple readers. I have multiple readers or 

Reviewer: ```suggestion
2. My stateful application is composed of leaders and followers or one writer and multiple readers. I have multiple readers or
```

   followers and my application can tolerate multiple pods going down at the same time. I want to update this application more than
   one pod at a time so that i get the new updates rolled out quickly, especially if the number of instances of my application are
   large.

In order to support such scenarios, in Kubernetes 1.24, we are releasing a new Alpha feature called MaxUnavailable for StatefulSets.This requires the feature flag `MaxUnavailableStatefulSet` to be enabled. Once enabled, you can specify a new field called
MaxUnavailable in RollingUpdate for STatefulSet, as follows:-

```
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: web
  namespace: default
spec:

>   podManagementPolicy: OrderedReady

Reviewer: ```suggestion
  podManagementPolicy: OrderedReady # you must set OrderedReady
```

web-2   0/1     Pending             0          0s
web-2   0/1     Pending             0          0s
web-2   0/1     ContainerCreating   0          0s
web-2   1/1     Running             0          1s
web-0   1/1     Terminating         0          6m6s ----> start terminating 0(only after 2 and 1 are running)
web-0   0/1     Terminating         0          6m7s
web-0   0/1     Pending             0          0s
web-0   0/1     Pending             0          0s
web-0   0/1     ContainerCreating   0          0s
web-0   1/1     Running             0          1s
```
Note that as soon as the rolling update starts, both 4 and 3(the two highest ordinal pods) start terminating at the same time. Pods 
with ordinal 4 and 3 may become ready at their own pace. As soon as both pods 4 and 3 are ready, pods 2 and 1 start terminating at the
same time. When pods 2 and 1 are both running and ready, pod 0 starts terminating. 

Note that, traditionally StatefulSets follow a strict ordering when updating pods and they would go from 4 to 3 to 2 and so on, one
pod at a time. When going one pod at a time, its not possible for 3 to be running and ready before 4. When `maxUnavailable` is more 

> than 1(in current example its 2), its possible that 3 becomes ready and running before 4 and thats ok. The application users setting

Reviewer: than 1 (in the example scenario I set `maxUnavailable` to 2), it iss possible that replica
3 becomes ready and running before replica 4 is ready—and that is ok.
If you're a developer and you set
``

> maxUnavailable to more than 1, should know that this is possible and their application should be able to handle such ordering issues, 

Reviewer: ```suggestion
`maxUnavailable` to more than 1, you should know that this outcome is possible and you
must ensure that your application is able to handle such ordering issues that occur
```

> if any. When maxUnavailable is more than 1, the ordering is guaranteed in between the batch of pods being updated. This means, 
> its guaranteed, that pods in batch 1(ordinal 2 and 1) cannot start updating, until pods in batch 0(ordinal 4 and 3) are ready.

Reviewer: Is this right?
```suggestion
if any. When you set `maxUnavailable` greatter than 1, the ordering is guaranteed in between
each batch of pods being updated. That guarantee means, that pods in update batch 2
(replicas 2 and 1) cannot start updating, until the pods in from batch 0 (replicas 4 and 3) are ready.

Although Kubernetes refers to these as _replicas_, your stateful application may have a different view -
you don't have to run an application where every stateful pod holds the same data as each other pod
in the same statefulset. The important thing here is that updates to StatefulSets happen in batches,
and you can now have a batch size larger than 1 (as an alpha feature).
```

> Also note, that the above behavior is with `PodManagementPolicy=OrderedReady`.

Reviewer: Should this be:
```suggestion
Also note, that the above behavior is with `podManagementPolicy: OrderedReady`. If you define a StatefulSet
as `podManagementPolicy: Parallel`, updates to all the Pods for that StatefulSet happen in parallel,
and not as a series of batches.
```

So, now you may have a lot of questions about:-

> - what is the behavior when PodManagementPolicy is Parallel
> - what is the behavior when Partition is non zero

Reviewer: ```suggestion

- what is the behavior when you set `podManagementPolicy: Parallel`?
- what is the behavior when you set `partition` to a value other than `0`
```

Well, why dont you try it out. This is an alpha feature, and the Kubernetes contributors are looking for feedback on this feature. Did
this help you achieve your stateful scenarios ? Did you find a bug or do you think the behavior as implemented is not intuitive or can

> break applications or catch them by surprise? Please open an issue [here](https://github.com/kubernetes/kubernetes/issues).

Reviewer: ```suggestion
break applications or catch them by surprise? Please
[open an issue](https://github.com/kubernetes/kubernetes/issues) to let us know.
```

Keep an eye on this space, for more blogs to dissect the behavior of this feature in the coming months.

> ***Reference***

Reviewer: ```suggestion

## Further reading and next steps {#next-steps}

- [Maximum unavailable Pods](/docs/concepts/workloads/controllers/statefulset/#maximum-unavailable-pods)
  in the StatefulSet documentation
```

- [KEP for MaxUnavailable for StatefulSet](https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/961-maxunavailable-for-statefulset)
- [Implementation](https://github.com/kubernetes/kubernetes/pull/82162/files)

> - [Feature Issue](https://github.com/kubernetes/enhancements/issues/961)

Reviewer: ```suggestion
- [Enhancement Tracking Issue](https://github.com/kubernetes/enhancements/issues/961)
```
